internal/testserver/datastore/postgres.go

```diff
@@ -19,69 +20,91 @@ import (
 	"github.com/authzed/spicedb/pkg/secrets"
 )
 
+const (
+	POSTGRES_TEST_USER            = "postgres"
+	POSTGRES_TEST_PASSWORD        = "secret"
+	POSTGRES_TEST_PORT            = "5432"
+	POSTGRES_TEST_MAX_CONNECTIONS = "500"
+	PGBOUNCER_TEST_PORT           = "6432"
+)
+
+type container struct {
+	hostHostname      string
+	hostPort          string
+	containerHostname string
+	containerPort     string
+}
+
 type postgresTester struct {
-	conn            *pgx.Conn
-	hostname        string
-	port            string
-	creds           string
-	targetMigration string
+	container
+	hostConn             *pgx.Conn
+	creds                string
+	targetMigration      string
+	pgbouncerProxy       *container
+	useContainerHostname bool
 }
 
 // RunPostgresForTesting returns a RunningEngineForTest for postgres
-func RunPostgresForTesting(t testing.TB, bridgeNetworkName string, targetMigration string, pgVersion string) RunningEngineForTest {
-	return RunPostgresForTestingWithCommitTimestamps(t, bridgeNetworkName, targetMigration, true, pgVersion)
+func RunPostgresForTesting(t testing.TB, bridgeNetworkName string, targetMigration string, pgVersion string, enablePgbouncer bool) RunningEngineForTest {
+	return RunPostgresForTestingWithCommitTimestamps(t, bridgeNetworkName, targetMigration, true, pgVersion, enablePgbouncer)
 }
 
-func RunPostgresForTestingWithCommitTimestamps(t testing.TB, bridgeNetworkName string, targetMigration string, withCommitTimestamps bool, pgVersion string) RunningEngineForTest {
+func RunPostgresForTestingWithCommitTimestamps(t testing.TB, bridgeNetworkName string, targetMigration string, withCommitTimestamps bool, pgVersion string, enablePgbouncer bool) RunningEngineForTest {
 	pool, err := dockertest.NewPool("")
 	require.NoError(t, err)
 
-	name := fmt.Sprintf("postgres-%s", uuid.New().String())
+	bridgeSupplied := bridgeNetworkName != ""
+	if enablePgbouncer && !bridgeSupplied {
+		// We will need a network bridge if we're running pgbouncer
+		bridgeNetworkName = createNetworkBridge(t, pool)
+	}
+
+	postgresContainerHostname := fmt.Sprintf("postgres-%s", uuid.New().String())
 
-	cmd := []string{"-N", "500"} // Max Connections
+	cmd := []string{"-N", POSTGRES_TEST_MAX_CONNECTIONS}
 	if withCommitTimestamps {
 		cmd = append(cmd, "-c", "track_commit_timestamp=1")
 	}
 
-	resource, err := pool.RunWithOptions(&dockertest.RunOptions{
-		Name:         name,
-		Repository:   "postgres",
-		Tag:          pgVersion,
-		Env:          []string{"POSTGRES_PASSWORD=secret", "POSTGRES_DB=defaultdb"},
-		ExposedPorts: []string{"5432/tcp"},
+	postgres, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Name:       postgresContainerHostname,
+		Repository: "postgres",
+		Tag:        pgVersion,
+		Env: []string{
+			"POSTGRES_USER=" + POSTGRES_TEST_USER,
+			"POSTGRES_PASSWORD=" + POSTGRES_TEST_PASSWORD,
+			// use md5 auth to align postgres and pgbouncer auth methods
+			"POSTGRES_HOST_AUTH_METHOD=md5",
+			"POSTGRES_INITDB_ARGS=--auth=md5",
```

Comment:
Newer postgres versions default to the `scram-sha-256` auth method. Pgbouncer defaults to `md5`. The auth methods need to align or the connection will fail. I went with md5 just because it was easier.

>  I can see this getting annoying when setting debug breakpoints in a test and trying to introspect the database.

If the concern is that you'll have to hash the password to connect then that's not the case. I've confirmed that I can easily connect to one of the running instances by just typing "secret" as the password.

listing running containers from tests
```
$ docker container ls
CONTAINER ID   IMAGE                      COMMAND                  CREATED         STATUS         PORTS                                              NAMES
c4d9cc7b0cd3   edoburu/pgbouncer:latest   "/entrypoint.sh /usr…"   2 seconds ago   Up 2 seconds   0.0.0.0:33099->5432/tcp, 0.0.0.0:33098->6432/tcp   pgbouncer-8d2d22c7-80ef-4285-947d-fa274d7797bb
7a7389a2d965   edoburu/pgbouncer:latest   "/entrypoint.sh /usr…"   2 seconds ago   Up 2 seconds   0.0.0.0:33097->5432/tcp, 0.0.0.0:33096->6432/tcp   pgbouncer-3d4c0775-93c7-492f-bc78-7501261850bf
87b72f88fdba   postgres:16                "docker-entrypoint.s…"   3 seconds ago   Up 2 seconds   0.0.0.0:33095->5432/tcp                            postgres-c2a4e67b-72ec-4049-a6dc-ccbf040dac6e
72f8d12b79e1   postgres:16                "docker-entrypoint.s…"   3 seconds ago   Up 2 seconds   0.0.0.0:33094->5432/tcp                            postgres-38289da4-2479-4e47-9178-ba6185ec7070
```


connecting via pgbouncer:
```
$ psql -h localhost -p 33096 -U postgres
Password for user postgres: <type secret>
psql (16.0)
Type "help" for help.

postgres=# 
```

connecting to postgres directly:
```
psql -h localhost -p 33094 -U postgres
Password for user postgres: <type secret>
psql (16.0)
Type "help" for help.

postgres=#
```

Reply:
I could look into setting the auth-method to `trust` if we want to remove passwords entirely from the tests
https://www.postgresql.org/docs/14/auth-pg-hba-conf.html

Reply:
@bradengroom 
>If the concern is that you'll have to hash the password to connect then that's not the case. I've confirmed that I can easily connect to one of the running instances by just typing "secret" as the password.

that's exactly it! I misunderstood what the flag did, so I'm totally fine with the change, does not affect debugging procedures.

>I could look into setting the auth-method to trust if we want to remove passwords entirely from the tests
https://www.postgresql.org/docs/14/auth-pg-hba-conf.html

I'd say that'd be a quality of life improvement for debugging purposes, but definitely something to do in a follow up, I'd leave that up to you 🙏🏻